daemon/algod/api/server/v2/handlers.go

```diff
@@ -1139,6 +1141,98 @@ func (v2 *Handlers) GetApplicationByID(ctx echo.Context, applicationID uint64) e
 	return ctx.JSON(http.StatusOK, response)
 }
 
+// GetApplicationBoxes returns the box names of an application
+// (GET /v2/applications/{application-id}/boxes)
+func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64, params generated.GetApplicationBoxesParams) error {
+	appIdx := basics.AppIndex(applicationID)
+	ledger := v2.Node.LedgerForAPI()
+	lastRound := ledger.Latest()
+	keyPrefix := logic.MakeBoxKey(appIdx, "")
+
+	// cast param.Max to zero, if param.Max points to some value, we keep the value
+	castedMax := nilToZero(params.Max)
+	maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication
+
+	// compute `maxReturnNum` from `castedMax` and `maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication`
+	// `maxReturnNum == 0` for returning everything
+	// - `castedMax == 0` and `maxBoxThreshold == 0`,
+	//       return everything
+	// - `castedMax == 0` and `maxBoxThreshold > 0`,
+	//       returns up to `maxBoxThreshold`, check if exceeds
+	// - `castedMax > 0` and `maxBoxThreshold == 0`,
+	//       return up to `castedMax`
+	// - `castedMax > 0` and `maxBoxThreshold > 0`,
+	//       return up to min(castedMax, maxBoxThreshold),
+	//       if return size is `maxBoxThreshold`, check if exceeds
+
+	var boxKeys []string
+	var err error
+
+	dominatedByQryParams := castedMax > 0 && (maxBoxThreshold >= castedMax || maxBoxThreshold == 0)
+	returnsAll := castedMax == 0 && maxBoxThreshold == 0
+
+	if dominatedByQryParams || returnsAll {
+		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, castedMax)
+		if err != nil {
+			return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
+		}
+	} else {
+		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, maxBoxThreshold+1)
```

Comment:
@tzaffi Let me know if the response addresses the question?

* The highlighted if-statement branch uses a `+ 1` in order to impose the max response limit check.  Since the search result count is not known before query execution, the invariant can only be checked post-query execution.
* The prior if-statement branch is either constrained below the response limit _or_ deliberately queries for all search results.

Reply:
@tzaffi I've opened a child PR that linearizes flow in an attempt to address the concerns you've raised:  https://github.com/algorand/go-algorand/pull/4543.

I'll mark ready for review once the build passes.  Feel welcomed to provide feedback beforehand.  We'll close the thread once the child PR reaches a conclusion.

Reply:
I'm looking at the code, thanks! I still befuddled by "+1". In the case that `maxBoxThreshold` is non-zero (which I'm assuming is the case that is bounded) there will be no need to add 1 to this already positive number to avoid unbounded behavior.

Reply:
Resolving as I finally see what's going on, and the conversation is better continued on #4543